Redefine OPAQUE composition and modes (#255)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

  `params` to strengthen the input `msg` against offline dictionary attacks.
  This function also needs to satisfy collision resistance.

## Key Recovery Method {#deps-keyrec}

> OPAQUE relies on a key recovery mechanism so that a client can engage in the AKE.
> The key recovery mechanism MUST enable deterministic recovery of authentication material
> compatible with the AKE given the OPRF output and additional data.
> This additional data can be encapsulated in a specific `Envelope` structure, saved on
> the server during registration, and provided to the client during login.

Reviewer: ```suggestion
OPAQUE relies on a key recovery mechanism for storing and recovering authentication
material on the server. The key recovery storage mechanism takes as input a private 
seed and outputs an envelope. The retrieval process takes as input a private seed
and envelope and outputs authentication material. The signatures for these 
functionalities are below:

- Store(private_seed): build and return an `Envelope` structure
- Recover(private_seed, envelope): recover and return the authentication material
for the AKE from the Envelope.

The key recovery mechanism MUST be compatible with the chosen AKE.
```

Simplifying?

Author: Yeah, much better, thanks!
I suggest a little tweak though, to avoid any confusion on where the recovery takes place:

```suggestion
OPAQUE relies on a key recovery mechanism for storing authentication
material on the server and recovering it on the client. The key recovery storage
mechanism takes as input a private seed and outputs an envelope. The retrieval
process takes as input a private seed and envelope and outputs authentication
material. The signatures for these functionalities are below:

- Store(private_seed): build and return an `Envelope` structure
- Recover(private_seed, envelope): recover and return the authentication material
for the AKE from the Envelope.

The key recovery mechanism MUST be compatible with the chosen AKE.
```

 Also, before _"The key recovery storage mechanism takes as input a private seed and outputs an envelope."_ we don't define earlier what an envelope is. Do you think we should or leave it like that?

Author: I re-added `and the client's public key.` to the `Store()` signature, as it's necessary

The key recovery mechanism must provide the two functions allowing for the construction
of that `Envelope` structure and key recovery from that same `Envelope` structure.

> 

Reviewer: ```suggestion
```

Author: typo?

Reviewer: It's no longer relevant with the new changes.

We define an API for these functions, but future specifications are free to adapt their
signatures according to their needs:

- CreateEnvelope(randomized_input, additional_data): build and return an `Envelope` structure
and the client's public key key.
- RecoverEnvelope(randomized_input, envelope): recover and return the authentication material
for the AKE.
 
Any key recovery mechanism used in OPAQUE MUST NOT invalidate the security proof of OPAQUE.
Moreover, a chosen method must be compatible with the chosen AKE. For example, the key
recovery mechanism specified in {{key-recovery}} directly recovers a private key from a seed,
and the cryptographic primitive in the AKE must therefore support such a possibility.

## Authenticated Key Exchange (AKE) Protocol {#deps-ake}

> OPAQUE additionally depends on an Authenticated Key Exchange (AKE) protocol. This AKE
> MUST hold within 3 messages and allow the initiating client to input it's private key
> in its second step. 

Reviewer: ```suggestion
OPAQUE additionally depends on a three-message Authenticated Key Exchange (AKE) protocol 
which satisfies the forward secrecy and KCI properties discussed in {{sec-considerations}}.
```

This specification defines one particular AKE based on 3DH; see {{ake-protocol}}.
3DH assumes a prime-order group as described in {{I-D.irtf-cfrg-voprf, Section 2.1}}.
We let `Npk` and `Nsk` denote the size of public and private keys, respectively,
used in the AKE.

> The AKE protocol must provide the following functions for the client:

Reviewer: Maybe we can rewrite this as, "The AKE protocol messages are constructed using the following functions:"

> The AKE protocol must provide the following functions for the client:
> - Start(): Initiate the AKE by producing message ke1.
> - ClientFinalize(client_identity, client_private_key, server_identity, server_public_key, ke2): upon
> reception of the server's response ke2, complete the protocol for the client and produce ke3.
> 
> The AKE protocol must provide the following functions for the server:
> 
> - Response(server_identity, server_private_key, client_identity, record.client_public_key, ke1, response):
> upon reception of a client's request ke1, engage in the AKE.
> - ServerFinish(ke3): upon reception of a client's final AKE message ke3, complete the protocol for
> the server.

Reviewer: ```suggestion
```

Let's move this stuff around in a separate change.

~~~

The rest of this document describes the details of these stages in detail.
{{client-material}} describes how client credential information is
generated, encoded, stored on the server on registration, and recovered on login. {{offline-phase}} describes the
first registration stage of the protocol, and {{online-phase}} describes the
second authentication stage of the protocol. {{configurations}} describes how
to instantiate OPAQUE using different cryptographic dependencies and parameters.

# Client Credential Storage and Key Recovery {#client-material}

OPAQUE makes use of a structure called `Envelope` to manage client credentials.
The client creates its `Envelope` on registration and sends it to the server for
storage. On every login, the server sends this `Envelope` to the client so it can
recover its key material for use in the AKE.

> OPAQUE allows for different key recovery mechanisms. See {{key-recovery}} for specifications.

Reviewer: ```suggestion
Future variants of OPAQUE may use different key recovery mechanisms. See {{key-recovery}} for details.
```

6. Output cleartext_credentials
~~~

> ## Key Recovery {#key-recovery}

Reviewer: This seems like it belongs with the dependency section up top, where other parameters and constants are defined.

Author: What do you think?

I think that referencing `DeriveAuthKeyPair()` here makes sense for that the following functions.

Author: Or maybe we could add a line to show that this fills the requirement of compatibility with the AKE

A key recovery mechanism specifies the structure and encoding of the `Envelope`,
as well as the functions creating this `Envelope` and recovering the authentication
material. 

The size of the envelope is denoted `Ne` and may vary between different mechanisms.
Specifications MUST specify its exact value. If applications implement
{{preventing-client-enumeration}}, they MUST use the same mechanism throughout their
lifecycle in order to avoid activity leaks due to mode switching.

> This specification uses a key recovery mechanism we call `internal`, and detailed in
> the following section. An other mechanism, called `called` is described in appendix
> {{alternate-key-recovery}}. Future specifications MAY introduce alternate methods and
> their `Envelope` contents.

Reviewer: ```suggestion
```

Steps:
1. envelope_nonce = random(Nn)
2. masking_key = Expand(randomized_pwd, "MaskingKey", Nh)
3. auth_key = Expand(randomized_pwd, concat(envelope_nonce, "AuthKey"), Nh)
4. export_key = Expand(randomized_pwd, concat(envelope_nonce, "ExportKey"), Nh)
5. seed = Expand(randomized_pwd, concat(envelope_nonce, "PrivateKey"), Nsk)

> 6. _, client_public_key = DeriveAuthKeyPair(seed)

Author: That's why I'm referring to it some lines above, l. 595

Steps:
1. auth_key = Expand(randomized_pwd, concat(envelope.nonce, "AuthKey"), Nh)
2. export_key = Expand(randomized_pwd, concat(envelope.nonce, "ExportKey", Nh)
3. seed = Expand(randomized_pwd, concat(envelope.nonce, "PrivateKey"), Nsk)
4. client_private_key, client_public_key = DeriveAuthKeyPair(seed)
5. cleartext_creds = CreateCleartextCredentials(server_public_key,
                      client_public_key, server_identity, client_identity)
6. expected_tag = MAC(auth_key, concat(envelope.nonce, inner_env, cleartext_creds))
7. If !ct_equal(envelope.auth_tag, expected_tag),
     raise EnvelopeRecoveryError
8. Output (client_private_key, export_key)
~~~

> #### AKE compatibility

Reviewer: Should this go in the AKE section? It seems kinda out of place here.

Author: Yes, and reference this bit from here 👍

#### Shared Secret Derivation

Each party generates own private and public keyshares in each session using the following function.

> - GenerateAuthKeyPair(): Return a randomly generated private and public key pair. This can be

Reviewer: Why is this here?

Author: Not sure why it ended up here. I want to define it in the AKE section, since this function is only used there, for the client and server to get their DH keys. I'll revert and work over that in a different PR

# Acknowledgments

The OPAQUE protocol and its analysis is joint work of the author with Stanislaw
Jarecki and Jiayu Xu. We are indebted to the OPAQUE reviewers during CFRG's
aPAKE selection process, particularly Julia Hesse and Bjorn Tackmann.
This draft has benefited from comments by multiple people. Special thanks
to Richard Barnes, Dan Brown, Eric Crockett, Paul Grubbs, Fredrik Kuivinen,
Payman Mohassel, Jason Resch, Greg Rubin, and Nick Sullivan.

# Alternate Key Recovery Mechanisms {#alternate-key-recovery}

> Client authentication material can be recovered in different ways. They MUST
> respect the requirements specified in {{deps-keyrec}}. However, they may
> augment the `Store()` and `Recover()` signatures for accommodation.

Reviewer: ```suggestion
Client authentication material can be stored and retrieved using different key
recovery mechanisms, provided these mechanisms adhere to the requirements 
specified in {{deps-keyrec}}. Any key recovery mechanism that encrypts data
in the envelope MUST use an authenticated encryption scheme with random
key-robustness (or key-committing). Deviating from the key-robustness 
requirement may open the protocol to attacks, e.g., {{LGR20}}.
This specification enforces this property by using a MAC over the envelope 
contents. 

We remark that export_key for authentication or encryption requires 
no special properties from the authentication or encryption schemes 
as long as export_key is used only after authentication material is successfully
recovered, i.e., after the MAC in RecoverCredentials passes verification.
```

> In the following we outline an alternative key recovery mechanism. 

Reviewer: ```suggestion
```

> ## Encrypted Private key {#keyrec-encrypted}
> 
> The original definition of OPAQUE allows applications to import existing keys
> for the client. Here, the OPRF output is used to derive an encryption key, use
> it to encrypt a generated or imported private key, and include the result in the
> envelope.
> 
> ~~~
> struct {
>   uint8 nonce[Nn];
>   uint8 encrypted_creds[Nsk];
>   uint8 auth_tag[Nm];
> } Envelope;
> ~~~
> nonce: A unique nonce of length `Nn` used to protect this Envelope.
> 
> encrypted_creds : Encrypted client_private_key. Authentication of this field is
> ensured with the `auth_tag` field in the envelope.
> 
> auth_tag: Authentication tag protecting the contents of the envelope, covering
> the envelope nonce, the encrypted private key, and `CleartextCredentials`.
> 
> The interface described in {{deps-keyrec}} must be adapted to import a private
> key.
> 
> The `encrypted_creds` value is generated with the following process:
> 
> ~~~
> 1. pad = Expand(randomized_pwd, concat(nonce, "Pad"), len(client_private_key))
> 2. encrypted_creds = xor(client_private_key, pad)
> ~~~
> 
> nonce: The envelope nonce. 
> 
> randomized_pwd: A MHF hardened OPRF seed.
> 
> For recovery, the same sequence applied on `encrypted_creds` instead of
> `client_private_key` yields the client's private key. The client's public key
> can be recovered using the private key to avoid sending it with the envelope.

Reviewer: ```suggestion
```

> ### Envelope Encryption {#envelope-encryption}
> 
> The analysis of OPAQUE from {{OPAQUE}} requires the authenticated encryption
> scheme used to produce the encrypted envelope to have a special property called
> random key-robustness (or key-committing). This specification enforces this
> property by utilizing encrypt-then-MAC in the construction of the envelope.
> There is no option to use another authenticated encryption scheme with this
> specification. (Deviating from the key-robustness requirement may open the
> protocol to attacks, e.g., {{LGR20}}.) We remark that export_key for
> authentication or encryption requires no special properties from the
> authentication or encryption schemes as long as export_key is used only after
> the envelope is validated, i.e., after the MAC in RecoverCredentials passes
> verification.

Reviewer: ```suggestion
```

# Acknowledgments

The OPAQUE protocol and its analysis is joint work of the author with Stanislaw
Jarecki and Jiayu Xu. We are indebted to the OPAQUE reviewers during CFRG's
aPAKE selection process, particularly Julia Hesse and Bjorn Tackmann.
This draft has benefited from comments by multiple people. Special thanks
to Richard Barnes, Dan Brown, Eric Crockett, Paul Grubbs, Fredrik Kuivinen,
Payman Mohassel, Jason Resch, Greg Rubin, and Nick Sullivan.

# Alternate Key Recovery Mechanisms {#alternate-key-recovery}

Client authentication material can be recovered in different ways. In fact, the original
definition of OPAQUE proposed to derive an encryption key from the OPRF output, use it
to encrypt a generated or imported private key, and include the result in the envelope.

In the following we show how a future specification can define that method. Other
methods MAY introduce alternate `InnerEnvelope` contents.

> ## External Mode {#external-mode}

Reviewer: I would drop all of this text entirely, and just say that it's possible to implement another form of Envelope with a different protocol version.

Author: I thought we said to move it to the appendix
